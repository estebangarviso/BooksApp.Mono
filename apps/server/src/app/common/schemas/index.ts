export * from './page-options.schema';
