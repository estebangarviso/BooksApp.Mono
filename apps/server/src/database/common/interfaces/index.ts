export * from './deleted-at.interface';
export * from './timestamp.interface';
