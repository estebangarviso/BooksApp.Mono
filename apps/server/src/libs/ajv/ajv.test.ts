/* eslint-disable perfectionist/sort-objects */
import { BadRequestException } from '@nestjs/common';
import { type OpenAPIObject } from '@nestjs/swagger';
import { Type } from '@sinclair/typebox';
import { describe, expect, test, vi } from 'vitest';
import {
	AjvDto,
	AjvIterableDto,
	registerDtoOpenApiSchemas,
} from './ajv.dto.ts';
import { AjvValidationPipe } from './ajv.pipe.ts';

describe(AjvValidationPipe, () => {
	// tests
	test('register DTO schemas', () => {
		const openApi = {} as OpenAPIObject;

		class Dto extends AjvDto({ id: Type.Number() }) {}
		Dto.registerOpenApi();
		registerDtoOpenApiSchemas(openApi);

		const { components } = openApi;

		expect(components).toBeDefined();
		expect(components?.schemas?.[Dto.name]).toBeDefined();
	});

	test('DTO generates JSON schema', () => {
		Date.now = vi
			.fn()
			.mockReturnValue(new Date('2025-01-01T00:00:00.000Z').getTime());
		const dto = AjvDto({
			// primitive values
			string: Type.String(),
			number: Type.Number(),
			boolean: Type.Boolean(),
			// empty types
			undefined: Type.Undefined(),
			null: Type.Null(),
			void: Type.Void(),
			// allows any value
			any: Type.Any(),
			unknown: Type.Unknown(),
			// allows no values
			never: Type.Never(),
			// coercion - TypeBox doesn't have direct coercion, these are just typed as their target types
			coerceDate: Type.Date(),
			coerceToString: Type.String(),
			coerceToBool: Type.Boolean(),
			// literal
			tuna: Type.Literal('tuna'),
			twelve: Type.Literal(12),
			tru: Type.Literal(true),
			// strings
			max: Type.String({ maxLength: 3 }),
			min: Type.String({ minLength: 10 }),
			length: Type.String({ minLength: 5, maxLength: 5 }),
			email: Type.String({ format: 'email' }),
			url: Type.String({ format: 'uri' }),
			emoji: Type.String({ pattern: String.raw`^[\p{Emoji}]+$` }),
			uuid: Type.String({ format: 'uuid' }),
			cuid: Type.String({ pattern: '^c[a-z0-9]{24}$' }),
			cuid2: Type.String({ pattern: '^[a-z][a-z0-9]{7,31}$' }),
			ulid: Type.String({ pattern: '^[0-9A-HJKMNP-TV-Z]{26}$' }),
			regex: Type.String({ pattern: '[a-c]' }),
			includes: Type.String({ pattern: '.*hello.*' }),
			startsWith: Type.String({ pattern: '^a.*' }),
			endsWith: Type.String({ pattern: '.*z$' }),
			ipv4: Type.String({ format: 'ipv4' }),
			ipv6: Type.String({ format: 'ipv6' }),
			multiple: Type.String({
				format: 'email',
				maxLength: 12,
				default: 'test@test.cl',
			}),
			// numbers
			gt: Type.Number({ exclusiveMinimum: 5 }),
			gte: Type.Number({ minimum: 5 }),
			lt: Type.Number({ exclusiveMaximum: 5 }),
			lte: Type.Number({ maximum: 5 }),
			int: Type.Integer(),
			positive: Type.Number({ minimum: 0, exclusiveMinimum: 0 }),
			nonnegative: Type.Number({ minimum: 0 }),
			negative: Type.Number({ maximum: 0, exclusiveMaximum: 0 }),
			nonpositive: Type.Number({ maximum: 0 }),
			multipleOf: Type.Number({ multipleOf: 5 }),
			// nan - TypeBox doesn't have direct NaN support, using a custom type
			nan: Type.Number({ const: Number.NaN }),
			// dates
			datetime: Type.String({ format: 'date-time' }),
			from: Type.Date({ minimum: new Date('1900-01-01').getTime() }),
			until: Type.Date({ maximum: Date.now() }),
			// enum
			enum: Type.Union([
				Type.Literal('Salmon'),
				Type.Literal('Tuna'),
				Type.Literal('Trout'),
			]),
			// default
			default: Type.String({ default: 'hello world' }),
			// nullable, optionals
			optional: Type.Optional(Type.Boolean()),
			nullable: Type.Union([Type.Boolean(), Type.Null()]),
			nullish: Type.Union([
				Type.Boolean(),
				Type.Null(),
				Type.Undefined(),
			]),
			// objects
			object: Type.Object({
				id: Type.Number({ minimum: 0, exclusiveMinimum: 0 }),
				name: Type.String(),
			}),
			// arrays
			array: Type.Array(Type.Number()),
			arrayNonEmpty: Type.Array(Type.Number(), { minItems: 1 }),
			arrayOf5: Type.Array(Type.Number(), { minItems: 5, maxItems: 5 }),
			arrayFrom5: Type.Array(Type.Number(), { minItems: 5 }),
			// tuples
			tuple: Type.Tuple([Type.String(), Type.Boolean()]),
			// union
			stringOrNumber: Type.Union([Type.String(), Type.Number()]),
			numberOrBool: Type.Union([Type.Number(), Type.Boolean()]),
			discriminated: Type.Union([
				Type.Object({
					status: Type.Literal('success'),
					data: Type.String(),
				}),
				Type.Object({
					status: Type.Literal('failed'),
					message: Type.String(),
				}),
			]),
			// records
			record: Type.Record(Type.String(), Type.Number()),
			lookup: Type.Record(Type.String({ maxLength: 3 }), Type.Boolean()),
			// intersection
			intersection: Type.Intersect([
				Type.Object({
					name: Type.String(),
				}),
				Type.Object({
					role: Type.String(),
				}),
			]),
		});

		expect(dto.schema).toBeDefined();
		// clear snapshot

		expect(dto.schema).toMatchSnapshot();
	});

	test('object DTO parses values on instantiation', () => {
		class Dto extends AjvDto({ id: Type.Number() }) {}

		const dto = new Dto();

		expect(dto).toBeDefined();
	});

	test('iterable (array) DTO parses values on instantiation', () => {
		class DtoIterable extends AjvIterableDto([Type.Number()]) {}

		const dto = new DtoIterable([1]);

		expect(dto[0]).toBe(1);
	});

	test('iterable (tuple) DTO parses values on instantiation', () => {
		class DtoIterable extends AjvIterableDto([
			Type.Number(),
			Type.Boolean(),
		]) {}
		const dto = new DtoIterable([1, true]);

		expect(dto[0]).toBe(1);
		expect(dto[1]).toBe(true);
		expect(dto).toBeDefined();
	});

	describe('validation pipe', () => {
		const _pipe = new AjvValidationPipe();
		const _dto = AjvDto({
			number: Type.Number(),
			string: Type.String(),
		});
		const metadata = { metatype: _dto } as any;

		test('parses correct input', () => {
			const input = { number: 1, string: 'str' };

			const result = _pipe.transform(input, metadata);

			expect(result).toStrictEqual(input);
		});

		test('throws BadRequestException on bad input', () => {
			const input = { id: 1 };

			expect(() => _pipe.transform(input, metadata)).toThrow(
				BadRequestException,
			);
		});

		test('when metatype is not AjvDTO input is bypassed', () => {
			const input = { id: 1 };

			const result = _pipe.transform(input, { metatype: {} } as any);

			expect(result).toStrictEqual(input);
		});
	});
});
