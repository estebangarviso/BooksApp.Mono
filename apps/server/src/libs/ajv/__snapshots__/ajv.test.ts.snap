// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AjvValidationPipe > DTO generates JSON schema 1`] = `
{
  "properties": {
    "any": {
      Symbol(TypeBox.Kind): "Any",
    },
    "array": {
      "items": {
        "type": "number",
        Symbol(TypeBox.Kind): "Number",
      },
      "type": "array",
      Symbol(TypeBox.Kind): "Array",
    },
    "arrayFrom5": {
      "items": {
        "type": "number",
        Symbol(TypeBox.Kind): "Number",
      },
      "minItems": 5,
      "type": "array",
      Symbol(TypeBox.Kind): "Array",
    },
    "arrayNonEmpty": {
      "items": {
        "type": "number",
        Symbol(TypeBox.Kind): "Number",
      },
      "minItems": 1,
      "type": "array",
      Symbol(TypeBox.Kind): "Array",
    },
    "arrayOf5": {
      "items": {
        "type": "number",
        Symbol(TypeBox.Kind): "Number",
      },
      "maxItems": 5,
      "minItems": 5,
      "type": "array",
      Symbol(TypeBox.Kind): "Array",
    },
    "boolean": {
      "type": "boolean",
      Symbol(TypeBox.Kind): "Boolean",
    },
    "coerceDate": {
      "type": "Date",
      Symbol(TypeBox.Kind): "Date",
    },
    "coerceToBool": {
      "type": "boolean",
      Symbol(TypeBox.Kind): "Boolean",
    },
    "coerceToString": {
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "cuid": {
      "pattern": "^c[a-z0-9]{24}$",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "cuid2": {
      "pattern": "^[a-z][a-z0-9]{7,31}$",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "datetime": {
      "format": "date-time",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "default": {
      "default": "hello world",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "discriminated": {
      "anyOf": [
        {
          "properties": {
            "data": {
              "type": "string",
              Symbol(TypeBox.Kind): "String",
            },
            "status": {
              "const": "success",
              "type": "string",
              Symbol(TypeBox.Kind): "Literal",
            },
          },
          "required": [
            "status",
            "data",
          ],
          "type": "object",
          Symbol(TypeBox.Kind): "Object",
        },
        {
          "properties": {
            "message": {
              "type": "string",
              Symbol(TypeBox.Kind): "String",
            },
            "status": {
              "const": "failed",
              "type": "string",
              Symbol(TypeBox.Kind): "Literal",
            },
          },
          "required": [
            "status",
            "message",
          ],
          "type": "object",
          Symbol(TypeBox.Kind): "Object",
        },
      ],
      Symbol(TypeBox.Kind): "Union",
    },
    "email": {
      "format": "email",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "emoji": {
      "pattern": "^[\\p{Emoji}]+$",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "endsWith": {
      "pattern": ".*z$",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "enum": {
      "anyOf": [
        {
          "const": "Salmon",
          "type": "string",
          Symbol(TypeBox.Kind): "Literal",
        },
        {
          "const": "Tuna",
          "type": "string",
          Symbol(TypeBox.Kind): "Literal",
        },
        {
          "const": "Trout",
          "type": "string",
          Symbol(TypeBox.Kind): "Literal",
        },
      ],
      Symbol(TypeBox.Kind): "Union",
    },
    "from": {
      "minimum": -2208988800000,
      "type": "Date",
      Symbol(TypeBox.Kind): "Date",
    },
    "gt": {
      "exclusiveMinimum": 5,
      "type": "number",
      Symbol(TypeBox.Kind): "Number",
    },
    "gte": {
      "minimum": 5,
      "type": "number",
      Symbol(TypeBox.Kind): "Number",
    },
    "includes": {
      "pattern": ".*hello.*",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "int": {
      "type": "integer",
      Symbol(TypeBox.Kind): "Integer",
    },
    "intersection": {
      "allOf": [
        {
          "properties": {
            "name": {
              "type": "string",
              Symbol(TypeBox.Kind): "String",
            },
          },
          "required": [
            "name",
          ],
          "type": "object",
          Symbol(TypeBox.Kind): "Object",
        },
        {
          "properties": {
            "role": {
              "type": "string",
              Symbol(TypeBox.Kind): "String",
            },
          },
          "required": [
            "role",
          ],
          "type": "object",
          Symbol(TypeBox.Kind): "Object",
        },
      ],
      "type": "object",
      Symbol(TypeBox.Kind): "Intersect",
    },
    "ipv4": {
      "format": "ipv4",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "ipv6": {
      "format": "ipv6",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "length": {
      "maxLength": 5,
      "minLength": 5,
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "lookup": {
      "patternProperties": {
        "^(.*)$": {
          "type": "boolean",
          Symbol(TypeBox.Kind): "Boolean",
        },
      },
      "type": "object",
      Symbol(TypeBox.Kind): "Record",
    },
    "lt": {
      "exclusiveMaximum": 5,
      "type": "number",
      Symbol(TypeBox.Kind): "Number",
    },
    "lte": {
      "maximum": 5,
      "type": "number",
      Symbol(TypeBox.Kind): "Number",
    },
    "max": {
      "maxLength": 3,
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "min": {
      "minLength": 10,
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "multiple": {
      "default": "test@test.cl",
      "format": "email",
      "maxLength": 12,
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "multipleOf": {
      "multipleOf": 5,
      "type": "number",
      Symbol(TypeBox.Kind): "Number",
    },
    "nan": {
      "const": NaN,
      "type": "number",
      Symbol(TypeBox.Kind): "Number",
    },
    "negative": {
      "exclusiveMaximum": 0,
      "maximum": 0,
      "type": "number",
      Symbol(TypeBox.Kind): "Number",
    },
    "never": {
      "not": {},
      Symbol(TypeBox.Kind): "Never",
    },
    "nonnegative": {
      "minimum": 0,
      "type": "number",
      Symbol(TypeBox.Kind): "Number",
    },
    "nonpositive": {
      "maximum": 0,
      "type": "number",
      Symbol(TypeBox.Kind): "Number",
    },
    "null": {
      "type": "null",
      Symbol(TypeBox.Kind): "Null",
    },
    "nullable": {
      "anyOf": [
        {
          "type": "boolean",
          Symbol(TypeBox.Kind): "Boolean",
        },
        {
          "type": "null",
          Symbol(TypeBox.Kind): "Null",
        },
      ],
      Symbol(TypeBox.Kind): "Union",
    },
    "nullish": {
      "anyOf": [
        {
          "type": "boolean",
          Symbol(TypeBox.Kind): "Boolean",
        },
        {
          "type": "null",
          Symbol(TypeBox.Kind): "Null",
        },
        {
          "type": "undefined",
          Symbol(TypeBox.Kind): "Undefined",
        },
      ],
      Symbol(TypeBox.Kind): "Union",
    },
    "number": {
      "type": "number",
      Symbol(TypeBox.Kind): "Number",
    },
    "numberOrBool": {
      "anyOf": [
        {
          "type": "number",
          Symbol(TypeBox.Kind): "Number",
        },
        {
          "type": "boolean",
          Symbol(TypeBox.Kind): "Boolean",
        },
      ],
      Symbol(TypeBox.Kind): "Union",
    },
    "object": {
      "properties": {
        "id": {
          "exclusiveMinimum": 0,
          "minimum": 0,
          "type": "number",
          Symbol(TypeBox.Kind): "Number",
        },
        "name": {
          "type": "string",
          Symbol(TypeBox.Kind): "String",
        },
      },
      "required": [
        "id",
        "name",
      ],
      "type": "object",
      Symbol(TypeBox.Kind): "Object",
    },
    "optional": {
      "type": "boolean",
      Symbol(TypeBox.Kind): "Boolean",
      Symbol(TypeBox.Optional): "Optional",
    },
    "positive": {
      "exclusiveMinimum": 0,
      "minimum": 0,
      "type": "number",
      Symbol(TypeBox.Kind): "Number",
    },
    "record": {
      "patternProperties": {
        "^(.*)$": {
          "type": "number",
          Symbol(TypeBox.Kind): "Number",
        },
      },
      "type": "object",
      Symbol(TypeBox.Kind): "Record",
    },
    "regex": {
      "pattern": "[a-c]",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "startsWith": {
      "pattern": "^a.*",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "string": {
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "stringOrNumber": {
      "anyOf": [
        {
          "type": "string",
          Symbol(TypeBox.Kind): "String",
        },
        {
          "type": "number",
          Symbol(TypeBox.Kind): "Number",
        },
      ],
      Symbol(TypeBox.Kind): "Union",
    },
    "tru": {
      "const": true,
      "type": "boolean",
      Symbol(TypeBox.Kind): "Literal",
    },
    "tuna": {
      "const": "tuna",
      "type": "string",
      Symbol(TypeBox.Kind): "Literal",
    },
    "tuple": {
      "additionalItems": false,
      "items": [
        {
          "type": "string",
          Symbol(TypeBox.Kind): "String",
        },
        {
          "type": "boolean",
          Symbol(TypeBox.Kind): "Boolean",
        },
      ],
      "maxItems": 2,
      "minItems": 2,
      "type": "array",
      Symbol(TypeBox.Kind): "Tuple",
    },
    "twelve": {
      "const": 12,
      "type": "number",
      Symbol(TypeBox.Kind): "Literal",
    },
    "ulid": {
      "pattern": "^[0-9A-HJKMNP-TV-Z]{26}$",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "undefined": {
      "type": "undefined",
      Symbol(TypeBox.Kind): "Undefined",
    },
    "unknown": {
      Symbol(TypeBox.Kind): "Unknown",
    },
    "until": {
      "maximum": 1735689600000,
      "type": "Date",
      Symbol(TypeBox.Kind): "Date",
    },
    "url": {
      "format": "uri",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "uuid": {
      "format": "uuid",
      "type": "string",
      Symbol(TypeBox.Kind): "String",
    },
    "void": {
      "type": "void",
      Symbol(TypeBox.Kind): "Void",
    },
  },
  "required": [
    "string",
    "number",
    "boolean",
    "undefined",
    "null",
    "void",
    "any",
    "unknown",
    "never",
    "coerceDate",
    "coerceToString",
    "coerceToBool",
    "tuna",
    "twelve",
    "tru",
    "max",
    "min",
    "length",
    "email",
    "url",
    "emoji",
    "uuid",
    "cuid",
    "cuid2",
    "ulid",
    "regex",
    "includes",
    "startsWith",
    "endsWith",
    "ipv4",
    "ipv6",
    "multiple",
    "gt",
    "gte",
    "lt",
    "lte",
    "int",
    "positive",
    "nonnegative",
    "negative",
    "nonpositive",
    "multipleOf",
    "nan",
    "datetime",
    "from",
    "until",
    "enum",
    "default",
    "nullable",
    "nullish",
    "object",
    "array",
    "arrayNonEmpty",
    "arrayOf5",
    "arrayFrom5",
    "tuple",
    "stringOrNumber",
    "numberOrBool",
    "discriminated",
    "record",
    "lookup",
    "intersection",
  ],
  "type": "object",
  Symbol(TypeBox.Kind): "Object",
}
`;
